### PR TITLE
pull file From device and push file to device

### PR DESCRIPTION
**Pull request by zhou9584**
image

{
    "setUp": [{
            "method": "pullFileFromDevice",
            "args": ["/sdcard/Movies/testfile/"]
        },
		{
            "method": "pushFileToDevice",
            "args": ["D:\\files\\OpenSource\\Copy.ps1","/sdcard/Movies/testfile/"]
        }
    ],
    "tearDown": [{
            "method": "pullFileFromDevice",
            "args": ["/sdcard/Movies/testfile2/"]
        }
    ]
}

### CODE REVIEW

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/ActionExecutor.java [hunk 1]
```diff
@@ -29,54 +29,59 @@ public class ActionExecutor {
...
-        todoActions.forEach(deviceAction -> doAction(deviceManager, deviceInfo, logger, deviceAction));
+        for (DeviceAction deviceAction : todoActions) {
+            try {
+                doAction(deviceManager, deviceInfo, logger, deviceAction);
+            } catch (InvocationTargetException | IllegalAccessException e) {
+                isSuccess = false;
+                logger.error("Execute action: fail", e);
+            } catch (HydraLabRuntimeException e) {
+                isSuccess = false;
+                logger.error("Convert action arg: fail", e);
```
Comment (hydraxman) [inline agent/src/main/java/com/microsoft/hydralab/agent/runner/ActionExecutor.java:52]: Merge the 2 catches.

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/ActionExecutor.java [hunk 2]
```diff
@@ -29,54 +29,59 @@ public class ActionExecutor {
     /**
      * the implementation of supported actions should not be overload
      */
-    private Set<String> actionTypes = Set.of("setProperty", "setDefaultLauncher", "backToHome", "changeGlobalSetting", "changeSystemSetting");
+    private Set<String> actionTypes = Set.of("setProperty", "setDefaultLauncher", "backToHome", "changeGlobalSetting", "changeSystemSetting", "pushFileToDevice", "pullFileFromDevice");
 
-    public void doActions(@NotNull DeviceManager deviceManager, @NotNull DeviceInfo deviceInfo, @NotNull Logger logger,
-                          @NotNull Map<String, List<DeviceAction>> actions, @NotNull String when) {
+    public boolean doActions(@NotNull DeviceManager deviceManager, @NotNull DeviceInfo deviceInfo, @NotNull Logger logger,
```
Comment (hydraxman) [inline agent/src/main/java/com/microsoft/hydralab/agent/runner/ActionExecutor.java:34]: Maybe we can return an exception or an exception list here, so we can know more detail about the execute status in caller level.

File: common/src/main/java/com/microsoft/hydralab/common/util/ADBOperateUtil.java [hunk 3]
```diff
@@ -192,23 +192,30 @@ public boolean uninstallApp(DeviceInfo deviceInfo, String packageName, Logger lo
         return true;
     }
 
-    public void pullFileToDir(DeviceInfo deviceInfo, String filePath, String pathOnDevice, Logger logger) {
+    public void pushFileToDevice(DeviceInfo deviceInfo, String pathOnAgent, String pathOnDevice, Logger logger) throws Exception {
```
Comment (hydraxman) [inline common/src/main/java/com/microsoft/hydralab/common/util/ADBOperateUtil.java:195]: Clarify the detailed thrown exception here.

File: common/src/main/java/com/microsoft/hydralab/common/util/ADBOperateUtil.java [hunk 4]
```diff
@@ -192,23 +192,30 @@ public boolean uninstallApp(DeviceInfo deviceInfo, String packageName, Logger lo
...
+        String comm = String.format("push %s %s", pathOnAgent, pathOnDevice);
+        Process process = executeDeviceCommandOnPC(deviceInfo, comm, logger);
+        CommandOutputReceiver err = new CommandOutputReceiver(process.getErrorStream(), logger);
+        CommandOutputReceiver out = new CommandOutputReceiver(process.getInputStream(), logger);
+        err.start();
+        out.start();
+        process.waitFor(60, TimeUnit.SECONDS);
+    }
+
+    public void pullFileToDir(DeviceInfo deviceInfo, String pathOnAgent, String pathOnDevice, Logger logger) throws Exception {
```
Comment (hydraxman) [inline common/src/main/java/com/microsoft/hydralab/common/util/ADBOperateUtil.java:208]: Throws the specific exceptions.

File: agent/src/main/java/com/microsoft/hydralab/agent/runner/TestRunner.java [hunk 5]
```diff
@@ -145,6 +146,21 @@ protected void setUp(DeviceInfo deviceInfo, TestTask testTask, TestRun testRun)
     protected abstract void run(DeviceInfo deviceInfo, TestTask testTask, TestRun testRun) throws Exception;
 
     protected void tearDown(DeviceInfo deviceInfo, TestTask testTask, TestRun testRun) {
+        //execute actions
+        if (testTask.getDeviceActions() != null) {
+            testRun.getLogger().info("Start executing tearDown actions.");
+            Exception e = actionExecutor.doActions(deviceManager, deviceInfo, testRun.getLogger(), testTask.getDeviceActions(), DeviceAction.When.TEAR_DOWN);
+            if (e != null) {
+                testRun.getLogger().error("Execute actions failed when tearDown!");
```
Comment (hydraxman) [inline agent/src/main/java/com/microsoft/hydralab/agent/runner/TestRunner.java:154]: testRun.getLogger().error("Execute actions failed when tearDown!",e);

**Review by hydraxman — DISMISSED** [review]

**Review by hydraxman — DISMISSED** [review]

**Review by olivershen-wow — APPROVED** [review]

**Pull request merged**